File: medium/322-coin-change/322-coin-change.ts
function coinChange(coins: number[], amount: number, memo = {}): number {
    if (amount < 0) return -1;
    if (amount === 0) return 0;
    if (memo[amount]) return memo[amount];
    
    let minCombo = -1
    for (const coin of coins) {
        const curr = coinChange(coins, amount - coin, memo);
        if (curr !== -1) {
            if (minCombo === -1 || minCombo > curr) {
                minCombo = curr + 1;
            }
        }
    }
    memo[amount] = minCombo;
    return minCombo;
};

File: medium/322-coin-change/322-coin-change.rb
# @param {Integer[]} coins
# @param {Integer} amount
# @return {Integer}
def coin_change(coins, amount)
    dp = [false] * (amount + 1)
    dp[0] = 0
    1.upto(amount) do |i|
        poss = []
        coins.each do |coin|
            poss << dp[i - coin] if coin <= i && dp[i - coin] 
        end
        dp[i] = poss.min + 1 if poss.size > 0
    end
    dp[-1] ? dp[-1] : -1
end